deepchem/models/tensorgraph/models/symmetry_function_regression.py

```diff
@@ -122,17 +123,82 @@ def __init__(self,
     """
     self.n_tasks = n_tasks
     self.max_atoms = max_atoms
-    self.n_feat = n_feat
     self.layer_structures = layer_structures
     self.atom_number_cases = atom_number_cases
     super(ANIRegression, self).__init__(**kwargs)
     self.build_graph()
+    self.grad = None
+
+  def save(self):
+    self.grad = None # recompute grad on restore
+    super(ANIRegression, self).save()
+
+  def build_grad(self):
+    self.grad = tf.gradients(self.outputs, self.atom_feats)
+
+  def compute_grad(self, dataset, batch_size=1):
+    with self._get_tf("Graph").as_default():
+      if not self.built:
+        self.build()
+      if not self.grad:
+        self.build_grad()
+
+      feed_dict = dict()
+      X = dataset.X
+      flags = np.sign(np.array(X[:batch_size, :, 0]))
+      feed_dict[self.atom_flags] = np.stack([flags]*self.max_atoms, axis=2)*\
+          np.stack([flags]*self.max_atoms, axis=1)
+      feed_dict[self.atom_numbers] = np.array(X[:batch_size, :, 0], dtype=int)
+      feed_dict[self.atom_feats] = np.array(X[:batch_size, :, :], dtype=float)
+      return self.session.run([self.grad], feed_dict=feed_dict)
+
+  def pred_one(self, X, atomic_nums):
+    num_atoms = atomic_nums.shape[0]
+    X = X.reshape((num_atoms, 3))
+    A = atomic_nums.reshape((atomic_nums.shape[0], 1))
+    Z = np.zeros((23, 4))
+    Z[:X.shape[0], 1:X.shape[1]+1] = X
+    Z[:A.shape[0], :A.shape[1]] = A
+    X = Z
+    dd = dc.data.NumpyDataset(np.array(X).reshape((1, 23, 4)), np.array(0), np.array(1))
+    return self.predict(dd)[0]
+
+  def grad_one(self, X, atomic_nums):
+    y = self.pred_one(X, atomic_nums)
+    num_atoms = atomic_nums.shape[0]
+    X = X.reshape((num_atoms, 3))
+    A = atomic_nums.reshape((atomic_nums.shape[0], 1))
+    Z = np.zeros((23, 4))
+    Z[:X.shape[0], 1:X.shape[1]+1] = X
+    Z[:A.shape[0], :A.shape[1]] = A
+    X = Z
+    inp = np.array(X).reshape((1, 23, 4))
+    dd = dc.data.NumpyDataset(inp, np.array([y]), np.array([1]))
+    res = self.compute_grad(dd)[0][0][0]
+    res = res[:num_atoms, 1:]
+    return res.reshape((num_atoms*3,))
+
+  def minimize_structure(self, X, atomic_nums):
```

Comment:
Very interested to see this one in action! Could you add a docstring here too?

Reply:
I'll add a notebook, and happy to add a docstring :)

Reply:
Awesome!